Bluetooth: Audio: Add missing media proxy write callbacks

asbjornsabo opened the pull request:
This PR adds missing write callbacks to the media proxy (they were forgotten when the media proxy was done).
Writes no longer have to use the read/notify callbacks.

asbjornsabo committed:
Bluetooth: Audio: Media - callback renaming

Rename read and notification callback functions in preparation for
adding write callbacks.

No functional changes.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media write callbacks

Add callback functions for writes to writable characteristics.
(Control point writes will be handled in separate commits.)

Modify code to use write callbacks (instead of receive callbacks) for
writes.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - use write callbacks in tests

Add write callbacks to media_controller BabbleSim test.
Modify existing write tests to use the write callbacks.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - split control point callbacks

Split the single media control point callback into one callback for
command send (control point writes) and one callback for command
results (control point notifications).

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - use control point callbacks in tests

Add split callbacks to BabbleSim test.
Modify existing tests to use both callbacks.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - split search control point callbacks

Split the single searchcontrol point callback into one callback for
search send (search control point writes) and one callback for search
result codes (search control point notifications).

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - use search control points callback in tests

Add split search control point callbacks to BabbleSim test.
Modify existing test to use both callbacks.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -358,10 +358,23 @@ struct media_proxy_ctrl_cbs {
 
 	/**
-	 * @brief Playback Speed callback
+	 * @brief Track Position write callback
+	 *
+	 * Called when the Track Position is written
```
The "read" callbacks have this text:

```
	 * @brief Track Position receive callback
	 *
	 * Called when the Track Position is read or changed
```

Which is common for both the client and server, right? 

Unless we are expecting to use this callback for when the server updates a value itself (like setting the track position by turning a knob), I think this should be made clear that is a client operation on a remote device.

asbjornsabo replied:
The write callbacks (as the receive callbacks) will be called for both local and remote players.
This is on purpose, to unify the behaviour.

Would it be clearer if the wording was changed to "when the Track Position is set"?
("set" is used for the commands to set values, locally and remote, while "write" may be interpreted as being a GATT write, i.e. a remote write.)

Thalley replied:
You can't use `set` for control point operations though, so I think it's better to just keep it as "write". 

But if this callback is for both local and remote players, then we can keep it as is.

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -494,24 +573,36 @@ struct media_proxy_ctrl_cbs {
+	void (*media_state_recv)(struct media_player *player, int err, uint8_t state);
+
+	/**
+	 * @brief Command send callback
+	 *
+	 * Called when a command has been sent
+	 * See also media_proxy_ctrl_command_send()
+	 *
+	 * @param player   Media player instance pointer
+	 * @param err      Error value. 0 on success, GATT error on positive value
+	 *                 or errno on negative value.
+	 * @param cmd      The command sent
+	 */
+	void (*command_send)(struct media_player *player, int err, struct mpl_cmd cmd);
```
I think we should modify the tense of the words here. 

Typically we have used past tense for callbacks, e.g. `connected`, `written`, `sent`, `received`, `updated`, etc. 

I guess this should be changed to `command_sent` and the write callbacks as `x_written`. 

The "read" callbacks can stay `recv` as that could be `received`

asbjornsabo replied:
This is the naming that was agreed upon in the discussion with @joerchan, and which no-one objected to.
The reasons to have "write" and "recv" were that those are the names used in other parts of the stack.

Thalley replied:
I completed missed @joerchan replies in that thread somehow. He is correct that we use `read`/`write` for callbacks in other places, but we also have `_written` in the `ots` API, and general (as per my examples), the general tense is past tense for callbacks. 

But if you and Joakim prefers `write/receive` then we'll just do that.

asbjornsabo replied:
Personally, I am fine with "_written" as well, but since I had asked for his opinion (I actually think you suggested that) I thought I should follow it ;-)

Thalley replied:
My preferrence is to use past-tense for callbacks, but it's fine as it then :) It's still an experimental API and can change later if we regret it.

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -344,10 +344,10 @@ struct media_proxy_ctrl_cbs {
 
 	/**
-	 * @brief Track Position callback
+	 * @brief Track Position receive callback
```
General comment: `receive` => `received`

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -412,10 +438,23 @@ struct media_proxy_ctrl_cbs {
+	 * @param player   Media player instance pointer
+	 * @param err      Error value. 0 on success, GATT error on positive value
+	 *                 or errno on negative value.
+	 * @param id	   The ID (48 bits) attempted to write
```
Bad indentation

Thalley replied:
Applies multiple places for `@param id`

Thalley replied:
Yes, please use spaces for indentation for places like here.

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -494,24 +573,36 @@ struct media_proxy_ctrl_cbs {
 	 * @param err      Error value. 0 on success, GATT error on positive value
 	 *                 or errno on negative value.
-	 * @param cmd_ntf  The result of the command
+	 * @param ntf      The result received
```
Suggest
```suggestion
	 * @param ntf      The notification received
```
or

```suggestion
	 * @param res      The result received
```

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/host/audio/media_proxy.h:
```diff
@@ -521,24 +612,41 @@ struct media_proxy_ctrl_cbs {
+	 *                      or errno on negative value.
+	 * @param result_code   Search result code
+	 */
+	void (*search_recv)(struct media_player *player, int err, uint8_t result_code);
```
Not sure if you have it on a TODO somewhere already, but please remember adding values for the `result_code` values in `media_proxy.h`. It currently seems like the implementation uses the `BT_MCS_OPC_NTF_` values.

asbjornsabo replied:
Will make a task for that.  It does require a bit more of changes, as some of the BT_MCS se are also CONFIG values.

asbjornsabo replied:
There are result code values in the `media_proxy.h`, though:
`/**
 * @brief Search operation result codes
 */
#define MEDIA_PROXY_SEARCH_SUCCESS  0x01
#define MEDIA_PROXY_SEARCH_FAILURE  0x02`

Thalley replied:
It seems to a bit of a mix. Some places use `MEDIA_PROXY_SEARCH_` but e.g. `media_proxy_pl_search_cb` uses the `BT_MCS_SCP_NTF_` values. 

I also see that my initial comment was incorrect, as the values I referred to was used for the `mpl_cmd_ntf`notifications, but are defined in `mcs.h`.

Thalley commented on subsys/bluetooth/shell/media_controller.c:
```diff
@@ -127,24 +127,44 @@ static void track_duration_cb(struct media_player *plr, int err, int32_t duratio
+static void playback_speed_recv_cb(struct media_player *plr, int err, int8_t speed)
+{
+	if (err) {
+		shell_error(ctx_shell, "Player: %p, Playback speed receive failed (%d)", plr, err);
```
Since you actually have added the attempted value in the callback, we might as well log those

Thalley replied:
The same applies multiple places

asbjornsabo replied:
Since this is the shell, the user will usually just have entered that value, so it normally does not matter much.
Anyway, this is already existing code, and out of scope fo rthis PR.

asbjornsabo committed:
Bluetooth: Audio: Media - whitespace fixes

Replaces tabs with spaces within Doxygen documentation.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Added missing calls to command_send callback and search_send callback in media_proxy.  Hopefully, that fixes the failing test.
Added on additional parameter to the search_send callback, for consistency.

Thalley reviewed (approved):
My comments have been addressed/fixed/discussed.
